src/auth.rs

```diff
@@ -0,0 +1,180 @@
+use crate::config::Config;
+use crate::error::Error;
+
+use reqwest::header::AUTHORIZATION;
+use rocket::{
+    fairing::{AdHoc, Fairing},
+    http::{Cookie, CookieJar, SameSite},
+    outcome::IntoOutcome,
+    request::{self, FromRequest, Request},
+    response::Redirect,
+};
+use rocket_oauth2::{OAuth2, TokenResponse};
+use serde::{Deserialize, Serialize};
+
+#[derive(Deserialize, Serialize)]
+pub struct LoginUrl {
+    pub login_url: String,
+}
+
+#[derive(Debug, strum_macros::EnumString)]
+pub enum OauthProvider {
+    Google,
+    Microsoft,
+}
+
+pub struct TokenCookie(String);
+
+impl std::fmt::Display for TokenCookie {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
+        write!(f, "{}", self.0)
+    }
+}
+
+#[rocket::async_trait]
+impl<'r> FromRequest<'r> for TokenCookie {
+    type Error = Error;
+
+    async fn from_request(request: &'r Request<'_>) -> request::Outcome<TokenCookie, Error> {
+        request
+            .cookies()
+            .get_private("token")
+            .and_then(|c| c.value().parse().ok())
+            .map(|t| TokenCookie(t))
+            .or_forward(())
+    }
+}
+
+struct Google;
+
+pub fn fairing_google() -> impl Fairing {
+    AdHoc::on_ignite("OAuth2", |rocket| async {
+        rocket
+            .mount("/", rocket::routes![login_google, redirect_google])
+            .attach(OAuth2::<Google>::fairing("google"))
+    })
+}
+
+#[rocket::get("/oauth/login?<redirect>")]
+fn login_google(redirect: String, oauth2: OAuth2<Google>, cookies: &CookieJar<'_>) -> Redirect {
```

Comment:
What exactly is this function supposed to do?

Reply:
This endpoint will redirect to the Google login page (using normal Oauth flow). The `redirect_` endpoints is where the user will end up getting redirected to from Google/Microsoft